### ISSUE TITLE
Wrong port with SIP TCP transport ?

### ISSUE THREAD

**Issue by renaud-sc** [resiprocate/resiprocate#320]
Hi,

I am using Windows. It seems that passing a custom port to addTransport when using TCP is not working. It woks fine with UDP. The assigned port seems to be in the  dynamic client port range (between 49152 and 65535).
The port I am willing to bind to arrives correctly in the InternalTransport::bind() function so don't undestand what's going wrong.

Does anyone have seen this before ?

Thanks

Cheers

**Comment by sgodin** [comment]
That is just the way TCP works.  The port you put in the addTransport call is the port resip will listen on for inbound TCP connections.  For outbound TCP connections an ephemeral port is used.  This allows multiple outbound TCP connections.

http://www.tcpipguide.com/free/t_TCPIPClientEphemeralPortsandClientServerApplicatio.htm

**Comment by sgodin** [comment]
There is currently no way to have the stack use TCP outbound connections without listening for inbound TCP connections.  Even if you are a client, you may need to accept inbound TCP connections for basic SIP RFC 3261 TCP operation.  Responses are always sent on the same connection, however new requests from the far end may come on a new inbound TCP connection.  The only time that's not needed is if you and your server implement RFC5626, or if the far-end ensures it always messages you on your inbound connection only (not just for response, but also for requests).

Your firewall rules would be setup for your servers/destination TCP port.

**Comment by renaud-sc** [comment]
Hi Scott,

Let me rephrase to be sure I understand correctly and other peoples arriving here would do as well.

A SIP client (eg softphone) doesn't always need to be registered to registrar to place calls and can establish direct P2P connections (provided they know how to reach the other peer) and therefore need to accept incoming/inbound connection. In UDP there is nothing special to be done as it's connection less and only one port is used to send and listen. That port can be configured with addTransport(). In TCP, a softphone need to be both a client (to reach a registrar or another peer) and a server to listen to inbound connection (for direct P2P calls for example). The port set with addTransport() is the listening port.

So it's not how "TCP works" it's more "how a SIP UA has to work in TCP mode".

Thanks for your help !
